module-system/sov-modules-api/src/gas.rs

```diff
@@ -0,0 +1,160 @@
+pub mod sov_modules_api {
+    use anyhow::Result;
+
+    use core::marker::PhantomData;
+
+    // We add `remainder_gas` to `WorkingSet` as a simple `u64`; the gas will already be converted
+    // to scalar value when charged from `WorkingSet`.
+
+    pub struct WorkingSet<C> {
+        remainder_gas: u64,
+        _context: PhantomData<C>,
+    }
+
+    impl<C> WorkingSet<C> {
+        pub fn new(remainder_gas: u64) -> Self {
+            Self {
+                remainder_gas,
+                _context: PhantomData,
+            }
+        }
+
+        pub fn charge_gas(&mut self, gas: u64) -> Result<()> {
+            self.remainder_gas = self
+                .remainder_gas
```

Comment:
here we could of self.remining_funds = self.self.remining_funds - gas.value(self.price)